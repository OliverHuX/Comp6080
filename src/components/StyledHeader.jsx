import React from 'react';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import Typography from '@material-ui/core/Typography';
import SchoolIcon from '@material-ui/icons/School';
import ExitToAppIcon from '@material-ui/icons/ExitToApp';
import SportsEsportsIcon from '@material-ui/icons/SportsEsports';
import { makeStyles } from '@material-ui/core/styles';
import logout from './logout';
import PropTypes from 'prop-types';

const useStyles = makeStyles((theme) => ({
  icon: {
    marginRight: theme.spacing(2),
  },
  logout: {
    position: 'relative',
    display: 'flex',
    marginLeft: 'auto',
    marginRight: 0,
    cursor: 'pointer'
  },
  icons: {
    position: 'relative',
    marginLeft: theme.spacing(2),
    marginRight: 0,
    cursor: 'pointer'
  },
  footer: {
    backgroundColor: theme.palette.background.paper,
    padding: theme.spacing(6),
  },
}));

const jumpto = () => {
  window.location.href = '/join'
}
export function IconButton ({ icon, handleOnClick }) {
  const classes = useStyles();
  if (!handleOnClick) handleOnClick = (icon === 'join' ? jumpto : logout)
  if (icon === 'join') {
    return (
      <SportsEsportsIcon className={classes.icons} align="right" onClick={handleOnClick}>
      </SportsEsportsIcon>
    )
  }
  return (
    <ExitToAppIcon className={classes.icons} align="right" onClick={handleOnClick}>
    </ExitToAppIcon>
  )
}
IconButton.propTypes = {
  icon: PropTypes.string,
  handleOnClick: PropTypes.func
};
export function StyledHeader ({ handleOnClick }) {
  const classes = useStyles();
  const token = localStorage.getItem('token');
  return (
    <AppBar position="relative">
      <Toolbar>
        <SchoolIcon className={classes.icon} />
        <Typography variant="h6" color="inherit" noWrap>
          BigBrain
        </Typography>
        <div className={classes.logout} color="inherit">
          <IconButton handleOnClick={handleOnClick} icon={'join'} />
          {token && (<IconButton handleOnClick={handleOnClick} icon={'logout'} />)}
        </div>
      </Toolbar>
    </AppBar>
  )
}
StyledHeader.propTypes = {
  handleOnClick: PropTypes.func
};
