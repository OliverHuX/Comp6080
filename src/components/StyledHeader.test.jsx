import React from 'react';
import { IconButton, StyledHeader } from './StyledHeader';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import Typography from '@material-ui/core/Typography';
import SportsEsportsIcon from '@material-ui/icons/SportsEsports';
import ExitToAppIcon from '@material-ui/icons/ExitToApp';

describe('IconButton Component - An icon with onClick handler', () => {
  const onClick = jest.fn();
  it('has the logout icon by defalt', () => {
    const button = shallow(<IconButton />);
    const target = <ExitToAppIcon></ExitToAppIcon>;
    expect(button.containsMatchingElement(target)).toBe(true);
  })
  it('has the join icon when icon is join', () => {
    const button = shallow(<IconButton icon={'join'} />);
    const target = <SportsEsportsIcon></SportsEsportsIcon>;
    expect(button.containsMatchingElement(target)).toBe(true);
  })
  it('triggers onClick event handler when clicked when icon is logout', () => {
    shallow(<IconButton handleOnClick={onClick}/>).simulate('click');
    expect(onClick).toHaveBeenCalledTimes(1);
  });
  it('triggers onClick event handler when clicked when icon is join', () => {
    shallow(<IconButton handleOnClick={onClick} icon={'join'} />).simulate('click');
    expect(onClick).toHaveBeenCalledTimes(1);
  });
  it('renders with minimal props (snapshot)', () => {
    const button = renderer.create(<IconButton/>).toJSON();
    expect(button).toMatchSnapshot();
  });
  it('renders with icon join props (snapshot)', () => {
    const button = renderer.create(<IconButton icon={'join'}/>).toJSON();
    expect(button).toMatchSnapshot();
  });
  it('renders with provided onClick handler (snapshot)', () => {
    const button = renderer.create(<IconButton handleOnClick={onClick}/>).toJSON();
    expect(button).toMatchSnapshot();
  });
  it('renders with provided onClick handler and icon join(snapshot)', () => {
    const button = renderer.create(<IconButton handleOnClick={onClick} icon={'join'}/>).toJSON();
    expect(button).toMatchSnapshot();
  });
});

describe('StyledHeader Component', () => {
  const onClick = jest.fn();
  it('has the title BigBrain by default', () => {
    const styledHeader = shallow(<StyledHeader />);
    const target = <Typography>BigBrain</Typography>;
    expect(styledHeader.dive().containsMatchingElement(target)).toBe(true);
  })
  it('has join icon by default', () => {
    const styledHeader = shallow(<StyledHeader />);
    expect(styledHeader.find(IconButton).length).toBe(1);
    const target = <IconButton icon={'join'}></IconButton>;
    expect(styledHeader.dive().containsMatchingElement(target)).toBe(true);
  })
  it('does not have the logout icon when not logged in by default', () => {
    const styledHeader = shallow(<StyledHeader />);
    expect(styledHeader.find(IconButton).length).toBe(1);
    const target = <IconButton icon={'logout'}></IconButton>;
    expect(styledHeader.dive().containsMatchingElement(target)).toBe(false);
  })
  it('triggers onClick event handler when join clicked ', () => {
    const styledHeader = shallow(<StyledHeader handleOnClick={onClick} />);
    const button = styledHeader.find(IconButton);
    expect(button.length).toBe(1);
    button.props().handleOnClick();
    expect(onClick).toHaveBeenCalledTimes(1);
  })
  it('has the logout icon when logged in', () => {
    localStorage.setItem('token', 'faskToken');
    const styledHeader = shallow(<StyledHeader />);
    expect(styledHeader.find(IconButton).length).toBe(2);
    const target = <IconButton icon={'logout'}></IconButton>;
    expect(styledHeader.dive().containsMatchingElement(target)).toBe(true);
  })
  it('renders with minimal props which has one logout button(snapshot)', () => {
    const styledHeader = renderer.create(<StyledHeader/>).toJSON();
    expect(styledHeader).toMatchSnapshot();
  });
  it('renders with provided onClick handler (snapshot)', () => {
    const styledHeader = renderer.create(<StyledHeader handleOnClick={onClick}/>).toJSON();
    expect(styledHeader).toMatchSnapshot();
  });
  it('renders with join button and provided onClick handler (snapshot)', () => {
    const styledHeader = renderer.create(<StyledHeader handleOnClick={onClick} icon={'join'} />).toJSON();
    expect(styledHeader).toMatchSnapshot();
  });
});
