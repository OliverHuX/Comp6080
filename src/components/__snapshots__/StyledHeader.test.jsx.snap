// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconButton Component - An icon with onClick handler renders with icon join props (snapshot) 1`] = `
<svg
  align="right"
  aria-hidden={true}
  className="MuiSvgIcon-root makeStyles-icons-3"
  focusable="false"
  onClick={[Function]}
  viewBox="0 0 24 24"
>
  <path
    d="M21.58 16.09l-1.09-7.66C20.21 6.46 18.52 5 16.53 5H7.47C5.48 5 3.79 6.46 3.51 8.43l-1.09 7.66C2.2 17.63 3.39 19 4.94 19c.68 0 1.32-.27 1.8-.75L9 16h6l2.25 2.25c.48.48 1.13.75 1.8.75 1.56 0 2.75-1.37 2.53-2.91zM11 11H9v2H8v-2H6v-1h2V8h1v2h2v1zm4-1c-.55 0-1-.45-1-1s.45-1 1-1 1 .45 1 1-.45 1-1 1zm2 3c-.55 0-1-.45-1-1s.45-1 1-1 1 .45 1 1-.45 1-1 1z"
  />
</svg>
`;

exports[`IconButton Component - An icon with onClick handler renders with minimal props (snapshot) 1`] = `
<svg
  align="right"
  aria-hidden={true}
  className="MuiSvgIcon-root makeStyles-icons-3"
  focusable="false"
  onClick={[Function]}
  viewBox="0 0 24 24"
>
  <path
    d="M10.09 15.59L11.5 17l5-5-5-5-1.41 1.41L12.67 11H3v2h9.67l-2.58 2.59zM19 3H5c-1.11 0-2 .9-2 2v4h2V5h14v14H5v-4H3v4c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
  />
</svg>
`;

exports[`IconButton Component - An icon with onClick handler renders with provided onClick handler (snapshot) 1`] = `
<svg
  align="right"
  aria-hidden={true}
  className="MuiSvgIcon-root makeStyles-icons-3"
  focusable="false"
  onClick={[MockFunction]}
  viewBox="0 0 24 24"
>
  <path
    d="M10.09 15.59L11.5 17l5-5-5-5-1.41 1.41L12.67 11H3v2h9.67l-2.58 2.59zM19 3H5c-1.11 0-2 .9-2 2v4h2V5h14v14H5v-4H3v4c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
  />
</svg>
`;

exports[`IconButton Component - An icon with onClick handler renders with provided onClick handler and icon join(snapshot) 1`] = `
<svg
  align="right"
  aria-hidden={true}
  className="MuiSvgIcon-root makeStyles-icons-3"
  focusable="false"
  onClick={[MockFunction]}
  viewBox="0 0 24 24"
>
  <path
    d="M21.58 16.09l-1.09-7.66C20.21 6.46 18.52 5 16.53 5H7.47C5.48 5 3.79 6.46 3.51 8.43l-1.09 7.66C2.2 17.63 3.39 19 4.94 19c.68 0 1.32-.27 1.8-.75L9 16h6l2.25 2.25c.48.48 1.13.75 1.8.75 1.56 0 2.75-1.37 2.53-2.91zM11 11H9v2H8v-2H6v-1h2V8h1v2h2v1zm4-1c-.55 0-1-.45-1-1s.45-1 1-1 1 .45 1 1-.45 1-1 1zm2 3c-.55 0-1-.45-1-1s.45-1 1-1 1 .45 1 1-.45 1-1 1z"
  />
</svg>
`;

exports[`StyledHeader Component renders with join button and provided onClick handler (snapshot) 1`] = `
<header
  className="MuiPaper-root MuiAppBar-root MuiAppBar-positionRelative MuiAppBar-colorPrimary MuiPaper-elevation4"
>
  <div
    className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root makeStyles-icon-1"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M5 13.18v4L12 21l7-3.82v-4L12 17l-7-3.82zM12 3L1 9l11 6 9-4.91V17h2V9L12 3z"
      />
    </svg>
    <h6
      className="MuiTypography-root MuiTypography-h6 MuiTypography-colorInherit MuiTypography-noWrap"
    >
      BigBrain
    </h6>
    <div
      className="makeStyles-logout-2"
      color="inherit"
    >
      <svg
        align="right"
        aria-hidden={true}
        className="MuiSvgIcon-root makeStyles-icons-3"
        focusable="false"
        onClick={[MockFunction]}
        viewBox="0 0 24 24"
      >
        <path
          d="M21.58 16.09l-1.09-7.66C20.21 6.46 18.52 5 16.53 5H7.47C5.48 5 3.79 6.46 3.51 8.43l-1.09 7.66C2.2 17.63 3.39 19 4.94 19c.68 0 1.32-.27 1.8-.75L9 16h6l2.25 2.25c.48.48 1.13.75 1.8.75 1.56 0 2.75-1.37 2.53-2.91zM11 11H9v2H8v-2H6v-1h2V8h1v2h2v1zm4-1c-.55 0-1-.45-1-1s.45-1 1-1 1 .45 1 1-.45 1-1 1zm2 3c-.55 0-1-.45-1-1s.45-1 1-1 1 .45 1 1-.45 1-1 1z"
        />
      </svg>
      <svg
        align="right"
        aria-hidden={true}
        className="MuiSvgIcon-root makeStyles-icons-3"
        focusable="false"
        onClick={[MockFunction]}
        viewBox="0 0 24 24"
      >
        <path
          d="M10.09 15.59L11.5 17l5-5-5-5-1.41 1.41L12.67 11H3v2h9.67l-2.58 2.59zM19 3H5c-1.11 0-2 .9-2 2v4h2V5h14v14H5v-4H3v4c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
        />
      </svg>
    </div>
  </div>
</header>
`;

exports[`StyledHeader Component renders with minimal props which has one logout button(snapshot) 1`] = `
<header
  className="MuiPaper-root MuiAppBar-root MuiAppBar-positionRelative MuiAppBar-colorPrimary MuiPaper-elevation4"
>
  <div
    className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root makeStyles-icon-1"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M5 13.18v4L12 21l7-3.82v-4L12 17l-7-3.82zM12 3L1 9l11 6 9-4.91V17h2V9L12 3z"
      />
    </svg>
    <h6
      className="MuiTypography-root MuiTypography-h6 MuiTypography-colorInherit MuiTypography-noWrap"
    >
      BigBrain
    </h6>
    <div
      className="makeStyles-logout-2"
      color="inherit"
    >
      <svg
        align="right"
        aria-hidden={true}
        className="MuiSvgIcon-root makeStyles-icons-3"
        focusable="false"
        onClick={[Function]}
        viewBox="0 0 24 24"
      >
        <path
          d="M21.58 16.09l-1.09-7.66C20.21 6.46 18.52 5 16.53 5H7.47C5.48 5 3.79 6.46 3.51 8.43l-1.09 7.66C2.2 17.63 3.39 19 4.94 19c.68 0 1.32-.27 1.8-.75L9 16h6l2.25 2.25c.48.48 1.13.75 1.8.75 1.56 0 2.75-1.37 2.53-2.91zM11 11H9v2H8v-2H6v-1h2V8h1v2h2v1zm4-1c-.55 0-1-.45-1-1s.45-1 1-1 1 .45 1 1-.45 1-1 1zm2 3c-.55 0-1-.45-1-1s.45-1 1-1 1 .45 1 1-.45 1-1 1z"
        />
      </svg>
      <svg
        align="right"
        aria-hidden={true}
        className="MuiSvgIcon-root makeStyles-icons-3"
        focusable="false"
        onClick={[Function]}
        viewBox="0 0 24 24"
      >
        <path
          d="M10.09 15.59L11.5 17l5-5-5-5-1.41 1.41L12.67 11H3v2h9.67l-2.58 2.59zM19 3H5c-1.11 0-2 .9-2 2v4h2V5h14v14H5v-4H3v4c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
        />
      </svg>
    </div>
  </div>
</header>
`;

exports[`StyledHeader Component renders with provided onClick handler (snapshot) 1`] = `
<header
  className="MuiPaper-root MuiAppBar-root MuiAppBar-positionRelative MuiAppBar-colorPrimary MuiPaper-elevation4"
>
  <div
    className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root makeStyles-icon-1"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M5 13.18v4L12 21l7-3.82v-4L12 17l-7-3.82zM12 3L1 9l11 6 9-4.91V17h2V9L12 3z"
      />
    </svg>
    <h6
      className="MuiTypography-root MuiTypography-h6 MuiTypography-colorInherit MuiTypography-noWrap"
    >
      BigBrain
    </h6>
    <div
      className="makeStyles-logout-2"
      color="inherit"
    >
      <svg
        align="right"
        aria-hidden={true}
        className="MuiSvgIcon-root makeStyles-icons-3"
        focusable="false"
        onClick={[MockFunction]}
        viewBox="0 0 24 24"
      >
        <path
          d="M21.58 16.09l-1.09-7.66C20.21 6.46 18.52 5 16.53 5H7.47C5.48 5 3.79 6.46 3.51 8.43l-1.09 7.66C2.2 17.63 3.39 19 4.94 19c.68 0 1.32-.27 1.8-.75L9 16h6l2.25 2.25c.48.48 1.13.75 1.8.75 1.56 0 2.75-1.37 2.53-2.91zM11 11H9v2H8v-2H6v-1h2V8h1v2h2v1zm4-1c-.55 0-1-.45-1-1s.45-1 1-1 1 .45 1 1-.45 1-1 1zm2 3c-.55 0-1-.45-1-1s.45-1 1-1 1 .45 1 1-.45 1-1 1z"
        />
      </svg>
      <svg
        align="right"
        aria-hidden={true}
        className="MuiSvgIcon-root makeStyles-icons-3"
        focusable="false"
        onClick={[MockFunction]}
        viewBox="0 0 24 24"
      >
        <path
          d="M10.09 15.59L11.5 17l5-5-5-5-1.41 1.41L12.67 11H3v2h9.67l-2.58 2.59zM19 3H5c-1.11 0-2 .9-2 2v4h2V5h14v14H5v-4H3v4c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
        />
      </svg>
    </div>
  </div>
</header>
`;
